### ISSUE TITLE
x-for doesn't support Set object

### ISSUE THREAD

**Issue by brenjt** [alpinejs/alpine#1114]
AlpineJS does not seem to work with the Set object. Here is a demo replicating the problem

https://codepen.io/brenjt/pen/dyORwbP

Is there a plan to add this support?

**Comment by KevinBatdorf** [comment]
You need to use `size` instead of `.length` on a Set. It doesn't seem to be triggering a mutation though and I'm not sure whether it can be monitored. I can maybe play around with it later.

If you want it working now, you can do this:

```html
<div x-data="{items: new Set()}">
    <template x-for="item of items">
        <div x-text="item"></div>
    </template>
    
    <button @click.prevent="items.add(items.size + 1);$el.__x.updateElements($el)">Add to Set</button>
</div>
```

**Comment by SimoTod** [comment]
Only literal objects and arrays are reactive, as far as I remember. It's to do with the salesforce membrane alpine uses under the scene.

**Comment by ryangjchandler** [comment]
Yeah, @SimoTod is right. The underlying reactivity library doesn't wrap any objects in proxies, other than `Array` and `Object`. 

@brenjt This might be something that Alpine supports in the future, but for now I'd say this issue can be closed.

**Comment by HugoDF** [comment]
> Yeah, @SimoTod is right. The underlying reactivity library doesn't wrap any objects in proxies, other than `Array` and `Object`.
> 
> @brenjt This might be something that Alpine supports in the future, but for now I'd say this issue can be closed.

If Caleb's still going to use `vue-reactivity`, it looks like we'll get support for Set/Map etc see the tests https://github.com/vuejs/vue-next/blob/master/packages/reactivity/__tests__/reactive.spec.ts#L48

Probably good to close for now, maybe worth @calebporzio adding it to the v3 roadmap? (as something we get by changing the reactivity library)

**Comment by loilo** [comment]
Note for future Googlers (like me): Since the underlying reactivity engine has been swapped out to use [@vue/reactivity](https://www.npmjs.com/package/@vue/reactivity) in Alpine v3, Sets and Maps now _are_ supported in `Alpine.reactive()` as well.

**Comment by loilo** [comment]
@pohlt You're absolutely right, it seems like neither `x-for` nor `$watch` do properly work with a `Set`. Weirdly enough, re-rendering properly works when the set is spread into an array: https://codepen.io/loilo/pen/KKoKzyV?editors=1001

**Comment by ernstki** [comment]
@loilo Yep, `x-for="item in [..items]"` is an effective workaround—thanks!

**Comment by mullema** [comment]
Spreading a Map/Set into an array does not seem very efficient. Are there plans to have Maps and Sets work properly with Alpinejs?

**Comment by CoutoDev** [comment]
> @pohlt You're absolutely right, it seems like neither `x-for` nor `$watch` do properly work with a `Set`. Weirdly enough, re-rendering properly works when the set is spread into an array: https://codepen.io/loilo/pen/KKoKzyV?editors=1001

Thanks for this workaround!
